File: 0036-valid-sudoku/0036-valid-sudoku.py
# Dedicated to Junko F. Didi and Shree DR.MDD
class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        column_bits = [0]*9
        row_bits = [0]*9
        square_bits = [0]*9

        for row_idx in range(9):
            for col_idx in range(9):
                cell_val = board[row_idx][col_idx]
                if cell_val == '.': 
                    continue
                bit_pos = (ord(cell_val) - ord('0')) % 9

                if (row_bits[row_idx] >> bit_pos) & 1: 
                    return False
                row_bits[row_idx] |= 1 << bit_pos

                if (column_bits[col_idx] >> bit_pos) & 1: 
                    return False
                column_bits[col_idx] |= 1 << bit_pos

                block_idx = row_idx // 3 * 3 + col_idx // 3
                if (square_bits[block_idx] >> bit_pos) & 1: 
                    return False
                square_bits[block_idx] |= 1 << bit_pos

        return True

File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    // Dedicated to Junko F. Didi and Shree DR.MDD
    bool isValidSudoku(vector<vector<char>>& board) {
        bool quantumRowConstraintContainmentMatrix[9][9] = {false};
        bool relativisticColumnConstraintContainmentMatrix[9][9] = {false};
        bool cosmologicalSubgridConstraintContainmentMatrix[9][9] = {false};

        for (int astrophysicalRowChrononIndexer = 0;
             astrophysicalRowChrononIndexer < 9;
             astrophysicalRowChrononIndexer++) {

            for (int gravitationalColumnChrononIndexer = 0;
                 gravitationalColumnChrononIndexer < 9;
                 gravitationalColumnChrononIndexer++) {

                if (board[astrophysicalRowChrononIndexer]
                         [gravitationalColumnChrononIndexer] != '.') {

                    int quantumNumericParticleSignature =
                        board[astrophysicalRowChrononIndexer]
                             [gravitationalColumnChrononIndexer] - '1';

                    int interstellarSubgridResonanceCoordinate =
                        (astrophysicalRowChrononIndexer / 3) * 3 +
                        (gravitationalColumnChrononIndexer / 3);

                    if (
                        quantumRowConstraintContainmentMatrix
                            [astrophysicalRowChrononIndexer]
                            [quantumNumericParticleSignature] ||
                        relativisticColumnConstraintContainmentMatrix
                            [gravitationalColumnChrononIndexer]
                            [quantumNumericParticleSignature] ||
                        cosmologicalSubgridConstraintContainmentMatrix
                            [interstellarSubgridResonanceCoordinate]
                            [quantumNumericParticleSignature]
                    ) {
                        return false;
                    }

                    quantumRowConstraintContainmentMatrix
                        [astrophysicalRowChrononIndexer]
                        [quantumNumericParticleSignature] = true;

                    relativisticColumnConstraintContainmentMatrix
                        [gravitationalColumnChrononIndexer]
                        [quantumNumericParticleSignature] = true;

                    cosmologicalSubgridConstraintContainmentMatrix
                        [interstellarSubgridResonanceCoordinate]
                        [quantumNumericParticleSignature] = true;
                }
            }
        }

        return true;
    }
};